File: src/main/java/org/tindalos/principle/domain/core/Package.java
package org.tindalos.principle.domain.core;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

import com.google.common.base.Optional;
import com.google.common.collect.Lists;
import com.google.common.collect.Sets;

public abstract class Package extends PackageScala {

	public Package(String name) {
		super(name);
	}

	public PackageReference getReference() {
		return reference();
	}

    public void insert(Package aPackage) {
        if (this.equals(aPackage)) {
            throw new PackageStructureBuildingException("Attempted to insert into itself " + this);
        } else if (this.doesNotContain(aPackage)) {
            throw new PackageStructureBuildingException("Attempted to insert " + aPackage + " into " + this);
        } else if (this.isDirectSuperPackageOf(aPackage)) {
            subPackages().add(aPackage);
        } else {
            insertIndirectSubPackage(aPackage);
        }
    }

	// it dies if there are cycles
	// through references, not through subPackages. transaitive too
	public Set<PackageReference> cumulatedDependencies(Map<PackageReference, Package> packageReferenceMap) {
		return cumulatedDependenciesAcc(packageReferenceMap, new HashSet<PackageReference>());
	}

	private Set<PackageReference> cumulatedDependenciesAcc(Map<PackageReference, Package> packageReferenceMap, Set<PackageReference> dependencies) {

		Set<PackageReference> accumulatedPackageReferences = this.accumulatedDirectPackageReferences();

		accumulatedPackageReferences.removeAll(dependencies);

		if (accumulatedPackageReferences.isEmpty()) {
			dependencies.remove(this.reference());
			return dependencies;
		} else {
			Set<PackageReference> result = Sets.newHashSet(accumulatedPackageReferences);
			for (PackageReference packageReference : accumulatedPackageReferences) {
				Package aPackage = packageReferenceMap.get(packageReference);
				assert aPackage != null : packageReference + " does not exist!";
				dependencies.add(packageReference);
				result.addAll(aPackage.cumulatedDependenciesAcc(packageReferenceMap, dependencies));
				result.remove(this.reference());
			}
			return result;
		}
	}


    private void insertIndirectSubPackage(Package aPackage) {
        String relativeNameOfDirectSubPackage = aPackage.firstPartOfRelativeNameTo(this);
        getSubPackageByRelativeName(relativeNameOfDirectSubPackage).insert(aPackage);
    }
    
	// all the references going out from this package
	public Set<PackageReference> accumulatedDirectPackageReferences() {
		Set<PackageReference> packageReferences = Sets.newHashSet();
		for (Package child : subPackages()) {
			packageReferences.addAll(child.accumulatedDirectPackageReferences());
		}
		packageReferences.addAll(getOwnPackageReferences());
		packageReferences.remove(this.reference());
		return packageReferences;
	}

	private Set<Package> accumulatedDirectlyReferredPackages(Map<PackageReference, Package> packageReferenceMap) {
		Set<PackageReference> packageReferences = accumulatedDirectPackageReferences();
		Set<Package> packages = Sets.newHashSet();
		for (PackageReference reference : packageReferences) {
			packages.add(packageReferenceMap.get(reference));
		}
		assert !packages.contains(this) : "self-references should be excluded";
		return packages;
	}
	
	public CyclesInSubgraph detectCycles(Map<PackageReference, Package> packageReferences) {
		return detectCyclesOnThePathFromHere(TraversedPackages.empty(), CyclesInSubgraph.empty(), packageReferences);
	}

	public Map<PackageReference, Package> toMap() {
		return toMap(new HashMap<PackageReference, Package>());
	}

	private CyclesInSubgraph detectCyclesOnThePathFromHere(
			TraversedPackages traversedPackages, 
			CyclesInSubgraph foundCycles,
			Map<PackageReference, Package> packageReferences) {

	    //enough cycles have been found already with this package
	    if (foundCycles.isBreakingPoint(this)) {
	        return foundCycles;
	    }
	    foundCycles.rememberPackageAsInvestigated(this);
		
		// if we just closed a cycle, add it to the found list then return
		Optional<List<PackageReference>> cycleCandidateEndingHere = findCycleCandidateEndingHere(traversedPackages);
		if (cycleCandidateEndingHere.isPresent()) {
			if (isValid(cycleCandidateEndingHere.get())) {
				foundCycles.add(new Cycle(ListConverter.convert(cycleCandidateEndingHere.get())));
			}
		} else {
			for (Package referencedPackage : this.accumulatedDirectlyReferredPackages(packageReferences)) {

				CyclesInSubgraph cyclesInSubgraph = referencedPackage.detectCyclesOnThePathFromHere(traversedPackages.add(this.getReference()), foundCycles, packageReferences);

				foundCycles.mergeIn(cyclesInSubgraph);
			}
		}
		//System.err.println("Cycles found so far: " + foundCycles.getCycles().size());
		return foundCycles;
	}


	private boolean isValid(List<PackageReference> cycleCandidate) {
		if (cycleCandidate.size() < 2) {
			return false;
		}
		return notEveryNodeUnderFirst(cycleCandidate);
	}

	private Optional<List<PackageReference>> findCycleCandidateEndingHere(TraversedPackages traversedPackages) {

		int indexOfThisPackage = indexInTraversedPath(ListConverter.convert(traversedPackages.packages()));
		if (indexOfThisPackage > -1) {

			List<PackageReference> cycleCandidate = ListConverter.convert(traversedPackages.from(indexOfThisPackage));

			return Optional.of(cycleCandidate);
		} else {
			return Optional.absent();
		}

	}

	private int indexInTraversedPath(List<PackageReference> traversedPackages) {
		int index = traversedPackages.indexOf(this.getReference());
		if (index != -1) {
			// simple case
			return index;
		}

		for (index = 0; index < traversedPackages.size() - 1; index++) {

			PackageReference possibleMatch = traversedPackages.get(index);
			if (possibleMatch.equals(this.getReference())) {
				return index;
			} else if (this.getReference().isDescendantOf(possibleMatch)
			// &&
			// !this.getReference().isDescendantOf(traversedPackages.get(index+1))
					&& notAllAreDescendantsOf(traversedPackages.subList(index + 1, traversedPackages.size()), possibleMatch))

			{
				return index;
			}

		}
		// System.err.println("Failed " + traversedPackages + " " + this);
		return -1;
	}

}

File: src/main/java/org/tindalos/principle/domain/core/Package.scala
package org.tindalos.principle.domain.core

import scala.collection.mutable.ListBuffer
import org.apache.commons.lang3.builder.HashCodeBuilder
import com.google.common.collect.Lists
import com.google.common.collect.Sets
import com.google.common.base.Optional

abstract class Package(val reference: PackageReference) {

  val _subPackages: ListBuffer[Package] = ListBuffer()

  def subPackages = _subPackages.toList
  
  def this(referenceName: String) = this(new PackageReference(referenceName))

  def isUnreferred(): Boolean
  def getMetrics(): Metrics
  def getOwnPackageReferences(): Set[PackageReference]
  def instability = getMetrics().instability
  def distance = getMetrics().distance
  def isIsolated() = getMetrics().afferentCoupling == 0 && getMetrics().efferentCoupling == 0

  def insert(aPackage: Package) {
    if (this.equals(aPackage)) {
      throw new PackageStructureBuildingException("Attempted to insert into itself " + this)
    } else if (doesNotContain(aPackage)) {
      throw new PackageStructureBuildingException("Attempted to insert " + aPackage + " into " + this)
    } else if (isDirectSuperPackageOf(aPackage)) {
      _subPackages += aPackage
    } else {
      insertIndirectSubPackage(aPackage)
    }
  }

  // all the references going out from this package
  def accumulatedDirectPackageReferences(): Set[PackageReference] =
    subPackages.flatMap(_.accumulatedDirectPackageReferences()).toSet.filterNot(_.equals(reference)).toSet ++: subPackages.flatMap(_.accumulatedDirectPackageReferences()).toSet.filterNot(_.equals(reference)).toSet ++: getOwnPackageReferences()

  protected def accumulatedDirectlyReferredPackages(packageReferenceMap: Map[PackageReference, Package]): Set[Package] =
    accumulatedDirectPackageReferences().map(packageReferenceMap.get(_).get)

  def toMap(): Map[PackageReference, Package] = toMap(scala.collection.mutable.Map[PackageReference, Package]())

  protected def toMap(accumulatingMap: scala.collection.mutable.Map[PackageReference, Package]): Map[PackageReference, Package] = {

    accumulatingMap.put(reference, this)
    subPackages.foreach(child => child.toMap(accumulatingMap))
    accumulatingMap.toMap
  }

  protected def createNew(name: String) = {
    new Package(name) {
      override def getOwnPackageReferences() = Set()
      override def getMetrics() = UndefinedMetrics
      override def isUnreferred() = true
      override def isIsolated() = true
    }
  }

  protected def getSubPackageByRelativeName(relativeName: String) = {

    subPackages.find(_.reference.equals(reference.child(relativeName))) match {
      case Some(subPackage) => subPackage
      case None =>
        val directSubPackage = createNew(reference.createChild(relativeName))
        _subPackages += directSubPackage
        directSubPackage
    }
  }

  protected def indexInTraversedPath(traversedPackages: List[PackageReference]) = {
    val index = traversedPackages.indexOf(reference)
    if (index != -1) index
    else {
      var matchFoundIndex: Option[Int] = None
      for (index <- 0 to traversedPackages.length - 1 if matchFoundIndex.isEmpty) {
        val possibleMatch = traversedPackages(index)
        if (possibleMatch.equals(reference)
          || (reference.isDescendantOf(possibleMatch)
            && notAllAreDescendantsOf(traversedPackages.slice(index + 1, traversedPackages.length), possibleMatch))) {
          matchFoundIndex = Some(index)
        }

      }
      // System.err.println("Failed " + traversedPackages + " " + this)
      matchFoundIndex match {
        case None => -1
        case Some(index) => index
      }
    }
  }

  def detectCycles(packageReferences: Map[PackageReference, Package]) =
    detectCyclesOnThePathFromHere(TraversedPackages.empty(), CyclesInSubgraph.empty(), packageReferences)

  // it dies if there are cycles
  // through references, not through subPackages. transaitive too
  def cumulatedDependencies(packageReferenceMap: Map[PackageReference, Package]) = cumulatedDependenciesAcc(packageReferenceMap, scala.collection.mutable.Set[PackageReference]())

  private def cumulatedDependenciesAcc(packageReferenceMap: Map[PackageReference, Package], dependencies: scala.collection.mutable.Set[PackageReference]): Set[PackageReference] = {

    val accumulatedPackageReferences = this.accumulatedDirectPackageReferences().filterNot(dependencies.contains(_))

    if (accumulatedPackageReferences.isEmpty) {
      dependencies.filterNot(_.equals(reference)).toSet
    } else {
      var result = accumulatedPackageReferences
      accumulatedPackageReferences.foreach({ packageReference =>
        dependencies.add(packageReference)
        result = result ++: (packageReferenceMap.get(packageReference).get.cumulatedDependenciesAcc(packageReferenceMap, dependencies))
        result = result.filterNot(_.equals(reference))
      })
      result
    }
  }

  protected def detectCyclesOnThePathFromHere(
    traversedPackages: TraversedPackages, 
    foundCycles: CyclesInSubgraph, 
    packageReferences: Map[PackageReference, Package]): CyclesInSubgraph = {

    //enough cycles have been found already with this package
    if (foundCycles.isBreakingPoint(this)) foundCycles
    else {
      foundCycles.rememberPackageAsInvestigated(this)

      // if we just closed a cycle, add it to the found list then return
      val cycleCandidateEndingHere = findCycleCandidateEndingHere(traversedPackages)
      if (cycleCandidateEndingHere.isDefined) {
        if (isValid(cycleCandidateEndingHere.get)) {
          foundCycles.add(new Cycle(cycleCandidateEndingHere.get))
        }
      } else {
        accumulatedDirectlyReferredPackages(packageReferences).foreach({ referencedPackage =>
        val cyclesInSubgraph = referencedPackage.detectCyclesOnThePathFromHere(traversedPackages.add(reference), foundCycles, packageReferences)
        foundCycles.mergeIn(cyclesInSubgraph)
        })
      }
      //System.err.println("Cycles found so far: " + foundCycles.getCycles().size())
      foundCycles
    }
  }
  private def findCycleCandidateEndingHere(traversedPackages: TraversedPackages): Option[List[PackageReference]] = {

    val indexOfThisPackage = indexInTraversedPath(traversedPackages.packages)
    if (indexOfThisPackage > -1) Some(traversedPackages.from(indexOfThisPackage))
    else None
  }

  protected def notAllAreDescendantsOf(packages: List[PackageReference], possibleAncestor: PackageReference) = packages.exists(!_.isDescendantOf(possibleAncestor))

  protected def isDirectSuperPackageOf(aPackage: Package) = reference.isDirectParentOf(aPackage.reference)

  protected def doesNotContain(aPackage: Package) = !aPackage.reference.pointsInside(reference)

  protected def firstPartOfRelativeNameTo(parentPackage: Package) = reference.firstPartOfRelativeNameTo(parentPackage.reference)

  protected def notEveryNodeUnderFirst(cycleCandidate: List[PackageReference]) = {
    val first = cycleCandidate.head
    cycleCandidate.tail.find(!_.isDescendantOf(first)) match {
      case None => first.equals(reference)
      case Some(_) => true
    }
  }

  protected def isValid(cycleCandidate: List[PackageReference]) =
    if (cycleCandidate.length < 2) false
    else notEveryNodeUnderFirst(cycleCandidate)

  protected def insertIndirectSubPackage(aPackage: Package) = {
    val relativeNameOfDirectSubPackage = aPackage.firstPartOfRelativeNameTo(this)
    getSubPackageByRelativeName(relativeNameOfDirectSubPackage).insert(aPackage)
  }

  override def equals(other: Any) = other.isInstanceOf[Package] && other.asInstanceOf[Package].reference.equals(reference)

  override def hashCode() = new HashCodeBuilder().append(reference).hashCode()

  override def toString() = reference.toString()
}

class TraversedPackages(val packages: List[PackageReference] = List()) {

  def add(reference: PackageReference) = new TraversedPackages(packages :+ reference)
  def from(index: Int) = packages.slice(index, packages.length)
}

object TraversedPackages {
  def empty() = new TraversedPackages()
}